prometheus/metrics.go

```diff
@@ -93,40 +95,48 @@ func getServiceMetricsAsync(api v1.API, namespace string, servicename string, du
 	groupingIn := joinLabels(byLabelsIn)
 	groupingOut := joinLabels(byLabelsOut)
 
-	fetchRateAsync := func(metricName string) (chan Metric, chan Metric) {
-		chin := make(chan Metric)
-		chout := make(chan Metric)
-		go fetchRateRange(api, metricName, labelsIn, groupingIn, bounds, rateInterval, chin)
-		go fetchRateRange(api, metricName, labelsOut, groupingOut, bounds, rateInterval, chout)
-		return chin, chout
-	}
+	var wg sync.WaitGroup
+
+	var requestCountIn, requestCountOut *Metric
+	var requestSizeIn, requestSizeOut, requestDurationIn, requestDurationOut, responseSizeIn, responseSizeOut Histogram
 
-	fetchHistoAsync := func(metricName string) (chan Histogram, chan Histogram) {
-		chin := make(chan Histogram)
-		chout := make(chan Histogram)
-		go fetchHistogramRange(api, metricName, labelsIn, groupingIn, bounds, rateInterval, chin)
-		go fetchHistogramRange(api, metricName, labelsOut, groupingOut, bounds, rateInterval, chout)
-		return chin, chout
+	fetchRateInOut := func(p8sFamilyName string, metricIn **Metric, metricOut **Metric) {
+		m := fetchRateRange(api, p8sFamilyName, labelsIn, groupingIn, bounds, rateInterval)
+		*metricIn = m
+		m = fetchRateRange(api, p8sFamilyName, labelsOut, groupingOut, bounds, rateInterval)
+		*metricOut = m
+		wg.Done()
```

Comment:
should this be a defer wg.Done() at the top of the function? (same comment for getchHistoInOut, below)

Reply:
Isn't it the same? I know I used `defer` above in the same file so it doesn't look very consistent, I can change that.
But to my understanding, `defer` is just a handy way to cover all return paths, I don't know if it's better to use it on small functions with 1 single return path like this.

Reply:
If something fails within this func, `defer` ensures that `wg.Done()` will be called, sort of a finally (or that's what i understood)

Reply:
@jotak it may not be 100% necessary but there are, I think, implicit benefits.  By being at the top it shows clearly that this function is expected to be run as a goroutine and is part of some synchronization.  It also is just consistent, follows a pattern, and protects against any unforeseen changes to the func body.  So, basically, I think it's just a a good idea.  If you prefer not to change just add a comment and I'll approve/merge.  Otherwise, just make the code change and I'll merge.

Reply:
I changed it :)